File: problems/wf/2021/c.py
N, M = map(int, input().split())

q = 2
ans_p = -1
ans_q = -1
while True:
    anysmall = False
    r = 1
    while r < q:
        p = q - r
        num = M * p
        den = q**N - (q-p)**N
        if num % den == 0:
            ans_p = p
            ans_q = q
        if den <= num:
            anysmall = True
            if num % den == 0:
                anysmall = False
                break
        else:
            break
        r += 1
    if not anysmall:
        break
    q += 1

if ans_q == -1:
    print("impossible")
else:
    print(p, q)

File: problems/wf/2021/c.cpp
#pragma GCC optimize("Ofast")
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

#define repx(i, a, b) for (int i = a; i < b; i++)
#define rep(i, n) repx(i, 0, n)
#define invrepx(i, a, b) for (int i = b - 1; i >= a; i--)
#define invrep(i, n) invrepx(i, 0, n)

const ll INF = ll(1e18) + 1;

ll N, M;
vector<ll> qpow;

ll mul(ll a, ll b) {
    __int128 c = (__int128)a * b;
    if (c > INF) return INF;
    return c;
}

ll add(ll a, ll b) {
    ll c = a + b;
    if (c > INF) return INF;
    return c;
}

ll getden(ll q, ll r) {
    qpow.resize(N);
    qpow[0] = 1;
    rep(i, N - 1) qpow[i + 1] = mul(qpow[i], q);
    ll sm = 0, rpow = 1;
    rep(i, N) {
        sm = add(sm, mul(rpow, qpow[N - 1 - i]));
        rpow = mul(rpow, r);
    }
    return sm;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N >> M;

    for (ll q = 2;; q++) {
        bool anysmall = false;
        for (ll r = 1; r < q; r++) {
            ll den = getden(q, r);
            if (den <= M) {
                anysmall = true;
                if (M % den == 0) {
                    cout << (q - r) << " " << q << endl;
                    return 0;
                }
            } else {
                break;
            }
        }
        if (!anysmall) break;
    }

    cout << "impossible" << endl;
}
